File: 108-ConvertSortedArraytoBinarySearchTree/108-ConvertSortedArraytoBinarySearchTree.java
/*

Given an array where elements are sorted in ascending order, convert it to a height balanced BST.

For this problem, a height-balanced binary tree is defined as a binary tree in which the depth of the two subtrees of every node never differ by more than 1.

Example:

Given the sorted array: [-10,-3,0,5,9],

One possible answer is: [0,-3,9,-10,null,5], which represents the following height balanced BST:

      0
     / \
   -3   9
   /   /
 -10  5

Link : https://leetcode.com/problems/convert-sorted-array-to-binary-search-tree/

Approach : Recursive from centre (binary search)

*/


/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode sortedArrayToBST(int[] nums) {
        
        if (nums.length==0)
            return null;
        
        return getSortedArrayToBST(nums, 0, nums.length-1);
        
    }
    
    public static TreeNode getSortedArrayToBST( int[] nums, int start, int last)
    {
        if (start>last)
            return null;
        
        int mid= (start+last)/2;
        TreeNode node = new TreeNode(nums[mid]);
        node.left= getSortedArrayToBST(nums, start, mid-1);
        node.right= getSortedArrayToBST(nums, mid+1, last);
        return node;
    }
}

File: 108-ConvertSortedArraytoBinarySearchTree/108-ConvertSortedArraytoBinarySearchTree.py
"""

108. Convert Sorted Array to Binary Search Tree

Given an integer array nums where the elements are sorted in ascending order, convert it to a

binary search tree.

Example 1:

Input: nums = [-10,-3,0,5,9]
Output: [0,-3,9,-10,null,5]
Explanation: [0,-10,5,null,-3,null,9] is also accepted:

Example 2:

Input: nums = [1,3]
Output: [3,1]
Explanation: [1,null,3] and [3,1] are both height-balanced BSTs.

Constraints:

    1 <= nums.length <= 104
    -104 <= nums[i] <= 104
    nums is sorted in a strictly increasing order.
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        def dfs(lo, hi):
            if lo > hi: return None
            mid = lo + (hi - lo) // 2
            node = TreeNode(nums[mid])
            node.left = dfs(lo, mid - 1)
            node.right = dfs(mid + 1, hi)
            return node
        return dfs(0, len(nums) - 1)
